server/src/main/java/io/druid/client/cache/MapCache.java

```diff
@@ -78,24 +96,13 @@ public CacheStats getStats()
   @Override
   public byte[] get(NamedKey key)
   {
-    final byte[] retVal;
-    synchronized (clearLock) {
-      retVal = baseMap.get(computeKey(getNamespaceId(key.namespace), key.key));
-    }
-    if (retVal == null) {
-      missCount.incrementAndGet();
-    } else {
-      hitCount.incrementAndGet();
-    }
-    return retVal;
+    return baseCache.getIfPresent(computeKey(namespaceId.getUnchecked(key.namespace), key.key));
   }
 
   @Override
   public void put(NamedKey key, byte[] value)
   {
-    synchronized (clearLock) {
-      baseMap.put(computeKey(getNamespaceId(key.namespace), key.key), value);
-    }
+    baseCache.put(computeKey(namespaceId.getUnchecked(key.namespace), key.key), value);
```

Comment:
To prevent race conditions on namespaceId, the namespaceId can be stored, then the asMap() representation of namespaceId can be used to check and see if the namespaceId after baseCache.put is the same as when get was first called. if it is, then things are safe, if it is not, then the cache just populated should be removed.